pkg/util/docker/event_stream.go

```diff
@@ -98,23 +123,59 @@ CONNECT:
 			case <-cancelChan:
 				cancel()
 				return
-			case msg := <-messages:
-				event, err := d.processContainerEvent(msg)
-				if err != nil {
-					log.Debugf("skipping event: %s", err)
-					continue
-				}
-				dataChan <- event
 			case err := <-errs:
 				if err == io.EOF {
 					// Silently ignore io.EOF that happens on http connection reset
-					log.Debug("got EOF, re-connecting")
+					log.Debug("Got EOF, re-connecting")
 				} else {
 					log.Warnf("error getting docker events: %s", err)
 				}
 				cancel()
 				continue CONNECT // Re-connect to docker
+			case msg := <-messages:
+				event, err := d.processContainerEvent(msg)
+				if err != nil {
+					log.Debugf("Skipping event: %s", err)
+					continue
+				}
+
+				badSubs := d.eventState.dispatch(event)
+				for _, sub := range badSubs {
+					d.UnsubscribeFromContainerEvents(sub.name)
+				}
 			}
 		}
 	}
 }
+
+func (e *eventStreamState) dispatch(event *ContainerEvent) []*eventSubscriber {
+	var badSubs []*eventSubscriber
+	e.RLock()
+	for _, sub := range e.subscribers {
+		err := e.sendEvent(event, sub)
+		if err != nil {
+			badSubs = append(badSubs, sub)
+		}
+	}
+	e.RUnlock()
+	return badSubs
+}
+
+func (s *eventStreamState) sendEvent(ev *ContainerEvent, sub *eventSubscriber) error {
```

Comment:
so we send a reference, that means consumers must not modify the object as it's shared. Might be worth commenting about that, or mentioning it in the doc for this system if we have some yet?

Reply:
updated the doc of the `SubscribeToContainerEvents` method